### PR TITLE
Added WatchClass command to notify users that a class is open

### PR DESCRIPTION
**Pull request by mr4tt**

### CODE REVIEW

**Comment by mr4tt** [comment]
email :)

**Comment by mr4tt** [comment]
every time you call me a clown, i will send you an email

File: src/commands/ucsd/DeleteClass.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import { GeneralUtilities } from "../../utilities";
+import { DataRegistry } from "../../DataRegistry";
+import { parseCourseSubjCode } from "../enroll-data/helpers/Helper";
+import { PostgresWatch } from "../../utilities/PostgresWatch";
+
+export default class DeleteClass extends BaseCommand {
+    public constructor() {
+        super ({
+            cmdCode: "DELETECLASS",
+            formalCommandName: "DeleteClass",
```
Comment (ewang2002) [inline src/commands/ucsd/DeleteClass.ts]:
```suggestion
            formalCommandName: "Delete Class",
```

File: src/commands/ucsd/DeleteClass.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            });
+
+            return -1;
+        }
+
+        await ctx.interaction.deferReply();
+
+        const searchRes = DataRegistry.COURSE_LISTING.find((x) => {
+            const subjCourseSplit = x.subjCourse.split("/");
+            // Case where we might have SUBJ1 NUM1/SUBJ2 NUM2/.../SUBJn NUMn
```
Comment (ewang2002) [inline src/commands/ucsd/DeleteClass.ts]: What about the case when you have `SUBJ NUM1/NUM2/NUM3`?

File: src/commands/ucsd/DeleteClass.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+                content: `Your input, \`${course}\`, is not correctly formatted. It should look like \`SUBJ XXX\`.`,
+                ephemeral: true,
+            });
+
+            return -1;
+        }
+
+        await ctx.interaction.deferReply();
+
+        const searchRes = DataRegistry.COURSE_LISTING.find((x) => {
```
Comment (ewang2002) [inline src/commands/ucsd/DeleteClass.ts]: I would not use the `COURSE_LISTING` to check if a course is available. I don't maintain the course listings as often as I'd like to. Just make a request to WebReg to see if the course is there.

File: src/commands/ucsd/WatchClass.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import { GeneralUtilities } from "../../utilities";
+import { DataRegistry } from "../../DataRegistry";
+import { parseCourseSubjCode } from "../enroll-data/helpers/Helper";
+import { PostgresWatch } from "../../utilities/PostgresWatch";
+
+export default class WatchClass extends BaseCommand {
+    public constructor() {
+        super ({
+            cmdCode: "WATCHCLASS",
+            formalCommandName: "WatchClass",
```
Comment (ewang2002) [inline src/commands/ucsd/WatchClass.ts]:
```suggestion
            formalCommandName: "Watch Class",
```

File: src/commands/ucsd/WatchClass.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+                ephemeral: true,
+            });
+
+            return -1;
+        }
+
+        await ctx.interaction.deferReply();
+
+        // check this 
+        const searchRes = DataRegistry.COURSE_LISTING.find((x) => {
```
Comment (ewang2002) [inline src/commands/ucsd/WatchClass.ts]: Same thing as last comment

File: src/utilities/GeneralUtilities.ts [hunk 6]
```diff
@@ -40,6 +40,7 @@ export namespace GeneralUtilities {
             return await func();
         }
         catch (e) {
+            console.log(e);
```
Comment (ewang2002) [inline src/utilities/GeneralUtilities.ts]: Use your logger, or remove this.

File: src/utilities/PostgresWatch.ts [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            // list of courses/users/channels that we could potentially alert on
+            const courses = await getAllAlertCourses();
+
+            // create a grouped map of course: object
+            const map: { [course: string]: CourseList } = {};
+            for (const c of courses) {
+                if (!(c.course in map)) {
+                    map[c.course] = [];
+                }
+                if (c) {
```
Comment (ewang2002) [inline src/utilities/PostgresWatch.ts:191]: Shouldn't you check if `c` is not null, and then check if `c.course` is not in the map?

File: src/utilities/PostgresWatch.ts [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            fields: [],
+        };
+    }
+
+    /**
+     * Every minute, search for alerts that need to be sent now. Send if found. 
+     */
+    export function loop() {
+        // call function every minute
+        const seconds = 60;
```
Comment (ewang2002) [inline src/utilities/PostgresWatch.ts]: Define the constant outside of this function. Use the naming convention for constants (e.g., `const SECONDS: number = 60;`)

**Review by ewang2002 — CHANGES_REQUESTED** [review]

**Comment by mr4tt** [comment]
you

**Comment by mr4tt** [comment]
are

**Comment by mr4tt** [comment]
a

**Comment by mr4tt** [comment]
big

**Comment by mr4tt** [comment]
huge

**Comment by mr4tt** [comment]
clown

**Review by ewang2002 — CHANGES_REQUESTED** [review]
ur a clown

**Review by ewang2002 — APPROVED** [review]

**Comment by 0AYU0** [comment]
attention.  a clown is approaching,

**Pull request merged**